platform/api/esx/api.go

```diff
@@ -0,0 +1,869 @@
+	if err != nil {
+		return fmt.Errorf("uploading disks to ResourcePool: %v", err)
+	}
+	_ = object.NewVirtualMachine(a.server.client.Client, *entity)
+
+	return nil
+}
+
+func (a *API) TerminateDevice(vmName string) error {
+	finder, _, _, _, _, err := a.server.GetServerDefaults()
+	if err != nil {
+		return fmt.Errorf("couldn't get server defaults: %v", err)
+	}
+
+	vm, err := finder.VirtualMachine(a.server.ctx, vmName)
+	if err != nil {
+		return fmt.Errorf("couldn't find VM: %v", err)
+	}
+
+	return a.DeleteDevice(vm)
+}
+
+func (a *API) DeleteDevice(vm *object.VirtualMachine) error {
+	task, err := vm.PowerOff(a.server.ctx)
+	if err != nil {
+		return fmt.Errorf("powering off vm: %v", err)
+	}
+
+	// We don't check for errors on this task because it will throw an error
+	// if the VM is already in a powered off state
+	_ = task.Wait(a.server.ctx)
+
+	task, err = vm.Destroy(a.server.ctx)
+	if err != nil {
+		return fmt.Errorf("destroying vm: %v", vm)
+	}
+
+	return task.Wait(a.server.ctx)
+}
+
+func (server *ESXServer) BuildCISR(name string, flagOVAPath string, finder *find.Finder, defaultNetwork object.NetworkReference, resourcePool *object.ResourcePool, datastore *object.Datastore) (*TapeArchive, *types.OvfCreateImportSpecResult, error) {
+	var nets []types.OvfNetworkMapping
+	nets = append(nets, types.OvfNetworkMapping{
+		Name:    "test",
+		Network: defaultNetwork.Reference(),
+	})
+
+	archive := &TapeArchive{flagOVAPath}
+	envelope, err := archive.ReadEnvelope("*.ovf")
+	if err != nil {
+		return archive, nil, fmt.Errorf("reading envelope: %v", err)
+	}
+
+	ovfHandler := object.NewOvfManager(server.client.Client)
+	cisp := types.OvfCreateImportSpecParams{
+		DiskProvisioning:   "",
+		EntityName:         name,
+		IpAllocationPolicy: "",
+		IpProtocol:         "",
+		OvfManagerCommonParams: types.OvfManagerCommonParams{
+			DeploymentOption: "",
+			Locale:           "US"},
+		PropertyMapping: []types.KeyValue{},
+		NetworkMapping:  NetworkMap(finder, envelope),
+	}
+
+	descriptor, err := archive.ReadOvf("*.ovf")
+	if err != nil {
+		return archive, nil, fmt.Errorf("reading ovf: %v", err)
+	}
+	cisr, err := ovfHandler.CreateImportSpec(server.ctx, string(descriptor), resourcePool, datastore, cisp)
+	if err != nil {
+		return archive, nil, err
+	}
+	if cisr.Error != nil {
+		return archive, nil, errors.New(cisr.Error[0].LocalizedMessage)
+	}
+	return archive, cisr, nil
+}
+
+func (server *ESXServer) GetServerDefaults() (*find.Finder, *object.Datacenter, *object.ResourcePool, *object.Datastore, object.NetworkReference, error) {
+	finder := find.NewFinder(server.client.Client, true)
+	datacenter, err := finder.DefaultDatacenter(server.ctx)
+	if err != nil {
+		return finder, nil, nil, nil, nil, err
+	}
+	finder.SetDatacenter(datacenter)
+	resourcePool, err := finder.DefaultResourcePool(server.ctx)
+	if err != nil {
+		return finder, datacenter, nil, nil, nil, err
```

Comment:
By convention, in Go, functions usually either succeed or fail completely.  Just return all `nil`s on error, or even better, return a struct with fields.